text/3678-final.md

```diff
@@ -0,0 +1,123 @@
+implementations of the trait use an identical method implementation. For
+instance, this may be an assumption required for correctness.
+
+This RFC allows restricting the implementation of trait methods.
+
+This mechanism also faciliates marker-like traits providing no implementable
+methods, such that implementers only choose whether to provide the trait and
+never how to implement it; the trait then provides all the method
+implementations.
+
+One example of a trait in the standard library benefiting from this:
+`Error::type_id`, which has thus far remained unstable because it's unsafe to
+override. This RFC would allow stabilizing that method so users can call it,
+without permitting reimplementation of it.
+
+# Explanation
+[explanation]: #explanation
+
+When defining a trait, the definition can annotate methods or associated
+functions to restrict whether implementations of the trait can define them. For
+instance:
+
+```rust
+trait MyTrait: Display {
+    final fn method(&self) {
+        println!("MyTrait::method: {self}");
+    }
+}
+```
+
+A method or associated function marked as `final` must have a default body.
+
+When implementing a trait, the compiler will emit an error if the
+implementation attempts to define any method or associated function marked as
+`final`, and will emit a suggestion to delete the implementation.
+
+In every other way, a `final` method or associated function acts identically to
+any other method or associated function, and can be invoked accordingly:
+
+```rust
+fn takes_mytrait(m: &impl MyTrait) {
+    m.method();
+}
+```
+
+Note that in some cases, the compiler might choose to avoid placing a `final`
+method in the trait's vtable, if the one-and-only implementation does not
+benefit from monomorphization.
+
+Note that removing a `final` restriction is always forwards-compatible.
+
+The keyword `final` has been reserved since Rust 1.0, so this feature can ship
+identically in all editions.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+As with any language feature, this adds more surface area to the language.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Rather than using `final`, we could use the `impl(visibility)` syntax from [RFC
+3323](https://rust-lang.github.io/rfcs/3323-restrictions.html). This would
+allow more flexibility (such as overriding a method within the crate but not
+outside the crate), and would be consistent with other uses of RFC 3323. On the
+other hand, such flexibility would come at the cost of additional complexity,
+and would be less familiar to people who have seen `final` in other languages.
+We can always add such syntax for the more general cases in the future if
+needed; see the future possibilities section.
+
+We could use `#[final]` rather than `final`. However, since we already have the
+`final` keyword reserved, using that keyword seems syntactically simpler than
+an attribute.
```

Comment:
The motivation here doesn't seem strong enough here to justify new syntax. As far as how this affects API consumers, it seems more like a logical continuation of `#[must_use]`, `#[non_exhaustive]`, `#[deprecated]`, or `#[diagnostic::*]` - and it doesn't seem to be more likely to be used.

Also - is `final` intuitive enough for anyone not coming with C++ or Java? I found the keyword somewhat unintuitive in C++ and I don't think consistency with those languages is worth enough to not do something like `#[no_override]`, which is much more obvious (similar to `non_exhaustive`).

Reply:
The `final` keyword is also used in Swift

Reply:
I've reworded this to avoid coming across as a rationale, rather than just a description of an alternative.